pkg/retriever/graphsyncretriever.go

```diff
@@ -131,29 +131,55 @@ func (r *graphsyncRetrieval) RetrieveFromCandidates(candidates []types.Retrieval
 	}
 	// start retrievals
 	queryStartTime := time.Now()
-	r.waitGroup.Add(len(candidates))
-	for _, candidate := range candidates {
-		candidate := candidate
-		go func() {
-			runRetrievalCandidate(ctx, r.GraphSyncRetriever, r.request, r.Client, retrieval, queryStartTime, candidate)
-			r.waitGroup.Done()
-		}()
-	}
+	r.waitGroup.Add(1)
+	go func() {
+		defer r.waitGroup.Done()
+		for {
+			hasCandidates, candidates, err := asyncCandidates.Next(ctx)
+			if !hasCandidates || err != nil {
+				return
+			}
+			for _, candidate := range candidates {
+				candidate := candidate
+				r.waitGroup.Add(1)
+				go func() {
+					defer r.waitGroup.Done()
+					runRetrievalCandidate(ctx, r.GraphSyncRetriever, r.request, r.Client, retrieval, queryStartTime, candidate)
+				}()
+			}
+		}
+	}()
 
-	return collectResults(ctx, retrieval, len(candidates), r.eventsCallback)
+	go func() {
+		r.waitGroup.Wait()
```

Comment:
We're being bothered by the race detector here. Caused by this: https://cs.opensource.google/go/go/+/refs/tags/go1.20.1:src/sync/waitgroup.go;l=110

From WaitGroup#Add docs: _"Note that calls with a positive delta that occur when the counter is zero must happen before a Wait."_

Because a we reuse a `graphsyncRetrieval` across multiple calls to `RetrieveFromAsyncCandidates`, I think the `Wait()` is probably misplaced here. Maybe we need a local waitGroup isolated to this call; there's still the inline comment on the struct field that this is only used for testing purposes, but that's not true anymore. Perhaps a local one for this call, but keep the struct one also in sync with the `Add` and `Done` calls?